runtime/src/main/java/org/corfudb/runtime/collections/StreamListenerResumePolicy.java

```diff
@@ -0,0 +1,103 @@
+        this.namespace = namespace;
+        this.streamTag = streamTag;
+        this.tablesOfInterest = tablesOfInterest;
+    }
+
+    public StreamListenerResumePolicy(CorfuStore store, String namespace, String streamTag) {
+        this(store, namespace, streamTag, null);
+    }
+
+    /**
+     * Implementation of re-subscription policy in case resuming subscription from last processed entry fails.
+     */
+    public abstract void subscribeOnResumeError();
+
+    public void resumeSubscription() {
+        if (lastProcessedEntryTs == null) {
+            log.warn("Last processed entry timestamp has not been set, failed to resume subscription." +
+                    " Default to fallback subscription.");
+            subscribeOnResumeError();
+            return;
+        }
+
+        try {
+            log.info("Resume subscription on [tag:{}] {}$[{}] from {}", streamTag, namespace, tablesOfInterest,
+                    lastProcessedEntryTs);
+
+            if (tablesOfInterest == null) {
+                store.subscribeListener(this, namespace, streamTag, lastProcessedEntryTs);
+            } else {
+                store.subscribeListener(this, namespace, streamTag, tablesOfInterest, lastProcessedEntryTs);
+            }
+        } catch (StreamingException e) {
+            log.error("Failed to resume subscription [tag:{}] from last processed entry {}. Re-subscribe based on " +
+                    "implemented fallback.", streamTag, lastProcessedEntryTs, e);
+            this.subscribeOnResumeError();
+        }
+    }
+
+    @Override
+    public void onNextEntry(CorfuStreamEntries results) {
+        onNext(results);
+        this.lastProcessedEntryTs = results.getTimestamp();
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        log.error("Exception caught during streaming processing. Re-subscribe this listener to latest timestamp. " +
+                "Error={}", throwable);
+
+        if (throwable instanceof UnreachableClusterException) {
+            try {
+                IRetry.build(IntervalRetry.class, () -> {
+                    try {
+                        log.info("Subscribe onError");
+                        resumeSubscription();
+                    } catch (UnreachableClusterException e) {
+                        log.error("Error while attempting to re-subscribe listener after onError.", e);
+                        throw new RetryNeededException();
+                    }
+                    return null;
+                }).run();
+            } catch (InterruptedException e) {
+                log.error("Unrecoverable exception when attempting to re-subscribe listener.", e);
+                throw new UnrecoverableCorfuInterruptedError(e);
+            }
+        }
+
+        // This listener assumes data loss cannot be afford, therefore, full sync is
+        // performed to recover the full state and immediately re-subscribe based on
+        resumeSubscription();
```

Comment:
Same here.

Reply:
@annym This method will get called, even the above retry succeed. I see you catch StreamingException in resumeSubscription, but it will have cost to validateSyncAddress and then log an error message, which might be confusing to users. What do you think?

Reply:
oh yes, this should be with an else, that's what you're saying right?

Reply:
yes, you are right.